ADR/00xx-component-dependencies.md

```diff
@@ -0,0 +1,171 @@
+  * In this case, the dependent images must be rebuilt to include the common parent image update in order to determine the actual effect of the proposed change.
+  * This is a many-to-one dependency. Many component images depend on one common parent image.
+* A user team has an **OLM operator**. When they propose an update to one of their operands with new code, they want to see if that’s going to break their operator. In order to be fully tested, a single rebuilt image needs to be included as a reference in a bundle image in order to be tested as a whole unit when deployed via OLM, i.e. for integration tests ([RHTAP-992](https://issues.redhat.com/browse/RHTAP-992)). A real example of this is in [gatekeeper](https://github.com/gatekeeper/gatekeeper-operator) where the operator repo contains both the [controller code](https://github.com/gatekeeper/gatekeeper-operator/blob/main/controllers/gatekeeper_controller.go) and the [bundle metadata](https://github.com/gatekeeper/gatekeeper-operator/blob/main/config/manifests/bases/gatekeeper-operator.clusterserviceversion.yaml), which need to be built into separate images([1](https://github.com/gatekeeper/gatekeeper-operator/blob/main/Dockerfile) and [2](https://github.com/gatekeeper/gatekeeper-operator/blob/main/bundle.Dockerfile)), separate Components in AppStudio.
+  * In this case, the bundle image must be rebuilt for the operand image update to be tested at all (assuming OLM is the strategy for deploying the operator).
+  * This is a one-to-many dependency. One operator bundle depends on many operand images.
+* A user team just has **two components that depend on each other in a functional way**. When they propose an update to one component, they want to submit a corresponding change to the second component, and have those tested together before merging both.
+  * In this case, on any given day, component B’s PRs may or may not have dependencies on PRs from component A. Not all changes are linked.
+
+In theory **any combination of the cases above** could be present in an app. They could have one common parent image, many operand images that depend on that, and one or more operator bundles which depend on those.
+
+Today, users work around how complicated it is to manage digests themselves by instead using *floating tags*, which have the benefit of being easy to use - no need to update - but have the problem of being unclear. It's not exactly clear what you're building against if you refer to it by tag. Potentially insecure. We want to make it easy for users to do better.
+
+## Decision
+
+### Interface changes
+
+**Declared Component Dependencies**: Introduce a new field on the `Component` resource called `depends on` that lets one Component declare that another Component **depends on it**.
+
+* The `depends on` field is a list.
```

Comment:
Is there something we can use (existing file, existing configuration in CPaaS or other system a team may be migrating from) to populate this dependency information during import/configuration? or does this represent brand new toil that users have to set up and maintain? 

'depends on' seems a little ambiguous to me. Are you saying that when I configure component A, and set 'depends on' to point to Component B, the created relationship is Component B is a dependent of A? With text of 'depends on', I'd expect that to be set from Component B, pointing to A, am I getting mixed up?

Reply:
Would that be a two way relationship in the CRs? or just one component knows the dependency? If we want to show the full picture in the UI, it might be helpful to have that information directly available from either Component CR.

Reply:
It might be interesting to think through what kind of rough flows or information we would want to display in the UI, and figure out if the proposed schema has a reasonable way to get that data efficiently. I definitely get not wanting to maintain the two way link, but wonder how that could impact our ability to contextualize and show that dependency relationship in the UI (or if it doesn't). If I look at Component B's details, should it show that Component A depends on it? Would we want to show the PRs for dependent images we created in Component A's commit visualization?

Reply:
The CPaaS configs have a `depends_on` representation, but I scripted an inversion of that relationship as I found that it was easier to comprehend and record a `nudges` relationship: https://github.com/arewm/project_rendering

It can be possible to internally however is deemed most appropriate, but from a user specification, _I_ didn't like that representation.

Reply:
I'm going to resolve this and leave it as an detail for the HAS team to sort out as owners of the Component CR.

cc @elsony and @johnmcollier for visibility.